data/settings.txt
# Файл с настройками для робота

# Настройки для MoveIt

handlePosition_x = -0.6
handlePosition_y = -0.2
handlePosition_z = 1.1
gripperPickHandle = 0.5

# Настройки для Tesseract
# Настройки визуализации
eigenVector3d = 1.0

# Начальное положение робота по умолчанию
# joint_start_pos_0 = 1.583660
# joint_start_pos_1 = -3.106403
# joint_start_pos_2 = 2.684593
# joint_start_pos_3 = -1.621857
# joint_start_pos_4 = -1.559460
# joint_start_pos_5 = 0.083117

# joint_start_pos_0 = -0.030000
# joint_start_pos_1 = -0.923000
# joint_start_pos_2 = 2.307000
# joint_start_pos_3 = -2.964372
# joint_start_pos_4 = -1.483593
# joint_start_pos_5 = -1.578000

# joint_start_pos_0 = 1.51731
# joint_start_pos_1 = -0.796949
# joint_start_pos_2 = 2.25895
# joint_start_pos_3 = -2.89963
# joint_start_pos_4 = -1.56047
# joint_start_pos_5 = 0.020617

### Робот под столом (тест 8 симуляция)
joint_start_pos_0 = 1.670050
joint_start_pos_1 = -0.923000
joint_start_pos_2 = 1.997900
joint_start_pos_3 = -4.200772
joint_start_pos_4 = -1.483593
joint_start_pos_5 = -1.578000


# Конечное положение робота по умолчанию
# joint_end_pos_0 = 1.578475
# joint_end_pos_1 = -2.311579
# joint_end_pos_2 = 1.266597
# joint_end_pos_3 = -1.640372
# joint_end_pos_4 = -1.557170
# joint_end_pos_5 = 0.078778

# joint_end_pos_0 = 0.149000
# joint_end_pos_1 = -1.519000
# joint_end_pos_2 = -1.384000
# joint_end_pos_3 = -1.646312
# joint_end_pos_4 = 1.519000
# joint_end_pos_5 = 0.149000

# Тест немного ближе к столу
# joint_end_pos_0 = 2.008491
# joint_end_pos_1 = -1.824498
# joint_end_pos_2 = 1.258161
# joint_end_pos_3 = -1.109452
# joint_end_pos_4 = -1.434897
# joint_end_pos_5 = 0.019922


# Тест близко к полу
# joint_end_pos_0 = 1.51731
# joint_end_pos_1 = -0.796949
# joint_end_pos_2 = 2.25895
# joint_end_pos_3 = -2.89963
# joint_end_pos_4 = -1.56047
# joint_end_pos_5 = 0.020617


### Робот над столом (тест 8 симуляция)
joint_end_pos_0 = 1.670050
joint_end_pos_1 = -1.541200
joint_end_pos_2 = 1.379700
joint_end_pos_3 = -2.964372
joint_end_pos_4 = -1.483593
joint_end_pos_5 = -1.578000


# Промежуточные точки
#1
joint_middle_pos_0 = 1.612235
joint_middle_pos_1 = -0.816569
joint_middle_pos_2 = 2.250976
joint_middle_pos_3 = -2.967063
joint_middle_pos_4 = -1.554412
joint_middle_pos_5 = 0.002781

#2
joint_middle2_pos_0 = 1.612175
joint_middle2_pos_1 = -1.964709
joint_middle2_pos_2 = 2.362315
joint_middle2_pos_3 = -2.021962
joint_middle2_pos_4 = -1.585328
joint_middle2_pos_5 = 0.002769

# Включать ли промежуточные точки
# false - нет, true - да
joint_middle_include = true

# IP робота
# robot_ip = 172.17.0.2
# robot_ip = 127.0.0.1
# robot_ip = 192.168.131.40

# Данные для тестовой отправки на UR5
joint_test_pos_0 = 0.0
joint_test_pos_1 = -0.06
joint_test_pos_2 = -2.72
joint_test_pos_3 = -0.34
joint_test_pos_4 = 0.0
joint_test_pos_5 = 0.0

# Разные настройки
# delay_loop_rate = 0.0

# Настройки для стола
table_length = 0.550
table_width = 2.000
table_height = 0.02
table_pos_x = 0
table_pos_y = 0.75
table_pos_z = 0.6


# Настройки для коробки
box_length = 0.07
box_width = 0.03
box_height = 0.01
box_pos_x = 0.93
box_pos_y = 0.02
box_pos_z = -0.14869


# Тип коллизии при настройке профиля для выполнения алгоритма
# Допустимые значения: SINGLE_TIMESTEP, DISCRETE_CONTINUOUS, CAST_CONTINUOUS

collision_cost_config_type = DISCRETE_CONTINUOUS
collision_constraint_config_type = DISCRETE_CONTINUOUS


# параметры для движения реального робота или в симуляции ur_sim
# максимальное значение ur_speed = 3.14, ur_acceleration = 40, ur_blend = 2
# ur_speed = 0.1
# ur_acceleration = 10.0
# ur_blend = 0.0


# коррекция положения после attach объекта
x_pos_correct = 0.002
y_pos_correct = 0.0
z_pos_correct = 0.16
x_orient_correct = -1.57
y_orient_correct = -1.6
z_orient_correct = 0

src/settings/include/settings/Settings.hpp
#pragma once

#include <boost/algorithm/string.hpp>
#include <ros/ros.h>
#include <iostream>
#include <stdio.h>
#include <string>
#include <fstream>

#ifndef settings
#define settings

// Настройки для MoveIt
#define robotDefaultPose "up"
#define robotDefaultLink "gripper_base_link"
#define robotDefaultJoint "gripper_finger1_joint"
#define gripperDefaultPose "open"
#define PLANNING_GROUP "manipulator"
#define GRIPPER_GROUP "gripper"
#define ROBOT_DESCRIPTION "robot_description"
#define M_PI 3.14159265358979323846

float handlePosition_x = -0.4;
float handlePosition_y = -0.2;
float handlePosition_z = 1.1;
float gripperPickHandle = 0.5;

class SettingsClass {
    public:
        void update();
};

// Перезаписать конфиги из файла
inline void SettingsClass::update() {
    std::ifstream infile("/workspace/data/settings.txt");
    std::string line;

    ROS_INFO("Settings start update...");

    while (std::getline(infile, line)) {
        std::string delimiter = "=";

        if (line.length() > 0 && line.at(0) != '#' && std::count(line.cbegin(), line.cend(), '=') == 1) {
            size_t pos = 0;
            std::string token;
            while ((pos = line.find(delimiter)) != std::string::npos) {
                token = line.substr(0, pos);
                line.erase(0, pos + delimiter.length());
            }

            boost::trim_left(token);
            boost::trim_right(token);
            boost::trim_left(line);
            boost::trim_right(line);

            if (strcmp("handlePosition_x", token.c_str()) == 0) {
                handlePosition_x = stof(line);
                ROS_INFO("Setting updated. New value handlePosition_x = %f", handlePosition_x);

            } else if (strcmp("handlePosition_y", token.c_str()) == 0) {
                handlePosition_y = stof(line);
                ROS_INFO("Setting updated. New value handlePosition_y = %f", handlePosition_y);

            } else if (strcmp("handlePosition_z", token.c_str()) == 0) {
                handlePosition_z = stof(line);
                ROS_INFO("Setting updated. New value handlePosition_z = %f", handlePosition_z);

            } else if (strcmp("gripperPickHandle", token.c_str()) == 0) {
                gripperPickHandle = stof(line);
                ROS_INFO("Setting updated. New value gripperPickHandle = %f", gripperPickHandle);

            }
        }
    }
}

#endif